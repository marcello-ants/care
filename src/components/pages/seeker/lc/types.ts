import { GeneralProviderProfile } from '@/types/seeker';

export interface ProviderProfile extends GeneralProviderProfile {
  numberOfReviews: number;
  qualities?: ProviderQualities;
  services?: ProviderServices;
  firstName: string;
}

export interface ProviderQualities {
  certifiedNursingAssistant?: boolean;
  certifiedRegisterNurse?: boolean;
  certifiedTeacher?: boolean;
  childDevelopmentAssociate?: boolean;
  comfortableWithPets?: boolean;
  cprTrained?: boolean;
  crn?: boolean;
  doesNotSmoke?: boolean;
  doula?: boolean;
  earlyChildDevelopmentCoursework?: boolean;
  earlyChildhoodEducation?: boolean;
  firstAidTraining?: boolean;
  nafccCertified?: boolean;
  ownTransportation?: boolean;
  specialNeedsCare?: boolean;
  trustlineCertifiedCalifornia?: boolean;
}

export interface ProviderServices {
  carpooling?: boolean;
  craftAssistance?: boolean;
  errands?: boolean;
  groceryShopping?: boolean;
  laundryAssistance?: boolean;
  lightHousekeeping?: boolean;
  mealPreparation?: boolean;
  swimmingSupervision?: boolean;
  travel?: boolean;
}
