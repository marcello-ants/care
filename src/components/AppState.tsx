import { useRouter } from 'next/router';
import { createContext, useReducer, useContext, ReactNode, useEffect } from 'react';
import { debounce } from 'lodash-es';
import { configureScope, captureException, Severity } from '@sentry/nextjs';
import {
  initialAppState,
  rootReducer,
  isStateVersionCompatible,
  doesStateHaveDefaultKeys,
  validateState,
} from '@/state';
import patchAppState from '@/state/patch';
import { initialState as flowInitialState } from '@/state/flow';
import logger from '@/lib/clientLogger';
import { ClientLogger } from '@care/utils';
import getVertical from '@/utilities/verticalUtils';
import { SessionStorageHelper } from '@/utilities/SessionStorageHelper';
import { trackPageViewInSiftScience } from '@/utilities/siftScienceHelper';
import { AppAction, AppDispatch, AppState } from '@/types/app';
import { StateCheckingResponse } from '@/types/common';
import {
  LOCAL_STORAGE_STATE_KEY,
  CZEN_BASE_PATH,
  FLOWS,
  POST_A_JOB_ROUTES,
  CLIENT_FEATURE_FLAGS,
} from '@/constants';
import { useFeatureFlags } from '@/components/FeatureFlagsContext';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';

type AppStateProviderProps = {
  children: ReactNode;
  initialStateOverride?: AppState;
  czenJSessionId?: string | undefined;
  referrerCookie?: string | undefined | null;
  userHasAccount?: boolean;
  memberUuid?: string;
};

function getStateFromStorageOrDefault(defaultState: AppState) {
  let stateFromStorageJson;
  const sessionStorageAvailable = SessionStorageHelper.isAvailable();

  if (sessionStorageAvailable) {
    stateFromStorageJson = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
  } else {
    logger.warn({ event: 'sessionStorageUnavailable' });
  }

  if (stateFromStorageJson) {
    const stateFromStorage = JSON.parse(stateFromStorageJson) as AppState;
    try {
      const patchedState = patchAppState(stateFromStorage, defaultState);

      if (isStateVersionCompatible(patchedState) && doesStateHaveDefaultKeys(patchedState)) {
        return patchedState;
      }
      logger.warn({ event: 'invalidStateFromLocalStorage', state: stateFromStorageJson });
    } catch (e) {
      captureException(e, { level: Severity.Warning });
    }
  }
  return defaultState;
}

function persistState(state: AppState) {
  SessionStorageHelper.setItem(LOCAL_STORAGE_STATE_KEY, JSON.stringify(state));
}

const debouncePersistState = debounce(persistState, 500, {
  leading: true,
  trailing: true,
  maxWait: 1000,
});

function reducerWrapper(state: AppState, action: AppAction) {
  const nextState = rootReducer(state, action);
  debouncePersistState(nextState);
  return nextState;
}

export function redirectIfInvalidState(state: AppState, pathname: string): StateCheckingResponse {
  if (
    typeof Intl === 'undefined' ||
    typeof Intl.NumberFormat === 'undefined' ||
    typeof Intl.DateTimeFormat === 'undefined'
  ) {
    // TODO: Remove this with SC-214
    logger.error({ event: 'IntlUnavailable', source: pathname, invalidState: state });
    return {
      redirectCalled: true,
      readableMessage: 'Safari 9 and lower is not currently supported',
    };
  }

  const isStateValid = validateState(state, pathname);
  const redirect = !isStateValid;

  if (redirect) {
    SessionStorageHelper.removeItem(LOCAL_STORAGE_STATE_KEY);

    window.location.assign(CZEN_BASE_PATH);
  }

  return { redirectCalled: redirect };
}

export function getFlowName(pathname: string, oldFlowName?: string): string {
  let newFlowName = Object.values(FLOWS).find((flow) => pathname.includes(flow.path))?.name;

  if (!newFlowName) {
    // TODO: this shouldn't be needed after SC-123 is implemented
    const isPostAJob = Object.values(POST_A_JOB_ROUTES).some((route) => pathname.includes(route));
    if (isPostAJob) {
      newFlowName = 'SEEKER_POST_A_JOB';
    }
  }

  // Don't allow shifting from in-facility to seeker flow name
  if (newFlowName === FLOWS.SEEKER.name && oldFlowName === FLOWS.SEEKER_IN_FACILITY.name) {
    return FLOWS.SEEKER_IN_FACILITY.name;
  }

  return newFlowName ?? '';
}

function useFlowDetection(dispatch: AppDispatch, oldFlowName: string) {
  const { pathname } = useRouter();
  const newFlowName = getFlowName(pathname, oldFlowName);
  const vertical = getVertical();
  ClientLogger.setLoggerStatics({ flow: newFlowName, vertical });

  useEffect(() => {
    dispatch({ type: 'setFlowName', flowName: newFlowName });

    configureScope((scope) => {
      scope.setTags({
        flow: newFlowName,
        vertical,
      });
    });
  }, [newFlowName]);
}

const AppStateContext = createContext<AppState>(initialAppState);
const AppDispatchContext = createContext<AppDispatch>(() =>
  // eslint-disable-next-line no-console
  console.warn(
    'Calling default dispatch is a no-op! You are probably consuming app dispatch in a component above AppStateProvider'
  )
);

export const AppStateProvider = ({
  children,
  initialStateOverride,
  czenJSessionId,
  referrerCookie,
  userHasAccount,
  memberUuid,
}: AppStateProviderProps) => {
  const hydratedInitialAppState = {
    ...initialAppState,
    flow: {
      ...flowInitialState,
      czenJSessionId,
      referrerCookie,
      userHasAccount,
      memberUuid,
    },
  };
  const featureFlags = useFeatureFlags();
  const sessionStorageRetrievalSuccessEvaluation =
    featureFlags?.flags[CLIENT_FEATURE_FLAGS.SESSION_STORAGE_RETRIEVAL_SUCCESS];
  const shouldMeasureSessionStorageRetrievalSuccess =
    sessionStorageRetrievalSuccessEvaluation?.variationIndex === 1;

  const router = useRouter();

  const logSessionStorageError = () => {
    logger.info({ event: 'sessionStorageStateUnavailable', path: router.pathname });
    AnalyticsHelper.logEvent({
      name: 'sessionStorageStateUnavailable',
    });
  };
  const launchDarklyWrappedGetStateFromStorageOrDefault = (defaultState: AppState) => {
    if (shouldMeasureSessionStorageRetrievalSuccess) {
      try {
        const sessionStorageState = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
        if (sessionStorageState && sessionStorageState.length > 0) {
          const stateFromStorage = JSON.parse(sessionStorageState);
          if (stateFromStorage) {
            logger.info({ event: 'sessionStorageStateAvailable', path: router.pathname });
          } else {
            logSessionStorageError();
          }
        } else {
          logSessionStorageError();
        }
      } catch (_e) {
        logSessionStorageError();
      }
    }
    return getStateFromStorageOrDefault(defaultState);
  };

  const [state, dispatch] = useReducer(
    reducerWrapper,
    initialStateOverride ?? hydratedInitialAppState,
    launchDarklyWrappedGetStateFromStorageOrDefault
  );
  useFlowDetection(dispatch, state.flow.flowName);
  const routePath = router.asPath?.split('?')[0];
  const stateCheckResult = redirectIfInvalidState(state, routePath);
  if (stateCheckResult.redirectCalled) {
    if (stateCheckResult.readableMessage) {
      return (
        <>
          <p>Redirecting to Member Home Page...</p>
          <p>{stateCheckResult.readableMessage}</p>
        </>
      );
    }
    return <></>;
  }

  useEffect(() => {
    trackPageViewInSiftScience(state);
  }, [routePath]);

  return (
    <AppDispatchContext.Provider value={dispatch}>
      <AppStateContext.Provider value={state}>{children}</AppStateContext.Provider>
    </AppDispatchContext.Provider>
  );
};

AppStateProvider.defaultProps = {
  initialStateOverride: undefined,
  czenJSessionId: undefined,
  referrerCookie: undefined,
  userHasAccount: false,
  memberUuid: undefined,
};

export const useAppState = () => useContext(AppStateContext);
export const useAppDispatch = () => useContext(AppDispatchContext);

export const useSeekerState = () => useContext(AppStateContext).seeker;
export const useProviderState = () => useContext(AppStateContext).provider;
export const useProviderCCState = () => useContext(AppStateContext).providerCC;
export const useFlowState = () => useContext(AppStateContext).flow;
export const useSeekerCCState = () => useContext(AppStateContext).seekerCC;
export const useSeekerHKState = () => useContext(AppStateContext).seekerHK;
export const useSeekerPCState = () => useContext(AppStateContext).seekerPC;
export const useSeekerTUState = () => useContext(AppStateContext).seekerTU;
export const useLtcgState = () => useContext(AppStateContext).ltcg;
