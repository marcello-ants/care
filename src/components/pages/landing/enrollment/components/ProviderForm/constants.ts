export const PROVIDER_SERVICES_DESKTOP = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const PROVIDER_SERVICES_MOBILE = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const DROPOFF_URLS = {
  CHILD_CARE: '/swf/ccOptimizedFlow',
  SENIOR_CARE: '/app/enrollment/provider/sc ',
  PET_CARE: '/swf/pcOptimizedFlow',
  HOUSEKEEPING: '/swf/hkOptimizedFlow',
  TUTORING: '/swf/tuOptimizedFlow',
  SPECIAL_NEEDS: '/swf/snOptimizedFlow',
  GENERIC: '/swf/cgOptimizedFlow',
  AU_PAIR: '',
  CARE_GIGS: '',
};
