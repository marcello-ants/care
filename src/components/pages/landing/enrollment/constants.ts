export const SITTER_FORM = {
  providerServices: {
    title: 'What Services do you offer?',
    id: 'providerService',
    options: [
      {
        value: 'CHILD_CARE',
        text: 'Babysitter',
      },
      {
        value: 'CHILD_CARE',
        text: 'Nanny',
      },
      {
        value: 'SPECIAL_NEEDS',
        text: 'Special Needs Provider',
      },
      {
        value: 'TUTORING',
        text: 'Tutor',
      },
      {
        value: 'TUTORING',
        text: 'Private Lesson Instructor',
      },
      {
        value: 'SENIOR_CARE',
        text: 'Senior Care Provider',
      },
      {
        value: 'PET_CARE',
        text: 'Pet Care Provider',
      },
      {
        value: 'HOUSEKEEPING',
        text: 'Housekeeper',
      },
      {
        value: 'GENERIC',
        text: 'Errands &amp; Odd Jobs Provider',
      },
    ],
  },
  providerEmail: {
    id: 'email',
    placeholder: 'Email',
  },
  providerPassword: {
    id: 'password',
    placeholder: 'Password',
  },
  providerZip: {
    id: 'providerZip',
    placeholder: 'ZIP',
  },
};

export const GENERIC_FORM = {
  howDidYouHearAboutUs: {
    title: 'How did you hear about us?',
    id: 'howDidYouHearAboutUs',
    options: [
      {
        value: 'FACEBOOK',
        text: 'Facebook',
      },
      {
        value: 'BANNER_AD',
        text: 'Banner Ad',
      },
      {
        value: 'TV_AD',
        text: 'TV Ad',
      },
      {
        value: 'OTHER_SOCIAL_MEDIA',
        text: 'Other Social Media (Twitter, Instagram, Pinterest, LinkedIn)',
      },
      {
        value: 'PRESS_COVERAGE',
        text: 'Press Coverage (News, Magazine, Blog)',
      },
      {
        value: 'FRIENDS_OR_FAMILY',
        text: 'Friends or Family',
      },
      {
        value: 'ONLINE_VIDEO',
        text: 'Online Video (Hulu, Youtube)',
      },
      {
        value: 'SEARCH_ENGINE',
        text: 'Search Engine (Google, Bing)',
      },
      {
        value: 'PARENTING_GROUP',
        text: 'Parenting Group or Forum',
      },
      {
        value: 'OTHER',
        text: 'Other',
      },
    ],
  },
};
