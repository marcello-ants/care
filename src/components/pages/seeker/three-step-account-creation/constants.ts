import * as yup from 'yup';
import { ServiceType } from '@/__generated__/globalTypes';
import { VerticalsAbbreviation } from '@/constants';

export const validationSchema = yup.object({
  firstName: yup
    .string()
    .min(2, 'First name needs to have at least two letters.')
    .required('First name is required'),
  lastName: yup
    .string()
    .min(2, 'Last name needs to have at least two letters.')
    .required('Last name is required'),
});

export const SERVICE_TYPE_BY_VERTICAL: { [key: string]: ServiceType } = {
  HK: ServiceType.HOUSEKEEPING,
  PC: ServiceType.PET_CARE,
  TU: ServiceType.TUTORING,
  CC: ServiceType.CHILD_CARE,
  DC: ServiceType.CHILD_CARE,
  IB: ServiceType.CHILD_CARE,
  SC: ServiceType.SENIOR_CARE,
};

export const ERROR_CREATING_ACCOUNT = 'An error occurred creating your account, please try again.';

export const FACEBOOK_VERTICALS_ALLOWANCE = ['CC'];
export const HDYHAU_VERTICALS_ALLOWANCE: VerticalsAbbreviation[] = ['HK', 'TU', 'PC', 'DC'];
