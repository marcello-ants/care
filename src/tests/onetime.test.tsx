import React from 'react';
import dayjs from 'dayjs';
import { useRouter } from 'next/router';
import { cloneDeep } from 'lodash-es';

import MockDate from 'mockdate';
import DayJSUtils from '@date-io/dayjs';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';

import setAmPmFormat from '@/tests/12HourFormatForSnapshot';
import { POST_A_JOB_ROUTES } from '../constants';
import OneTimePage from '../pages/onetime';
import { AppStateProvider } from '../components/AppState';
import { initialAppState } from '../state';
import { AppState } from '../types/app';

jest.mock('next/router', () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

const initialAppStateClone = cloneDeep(initialAppState);
const initialStateOverrideWithQueryParamData: AppState = {
  ...initialAppStateClone,
  seeker: {
    ...initialAppStateClone.seeker,
    jobPost: {
      ...initialAppStateClone.seeker.jobPost,
      zip: '10004',
      servicesNeeded: [],
      intent: 'NOW',
      selfNeedsCare: false,
      careFrequency: 'onetime',
    },
  },
};

let mockRouter = null;

const startDatePlaceholder = 'Estimated start date';
const endDatePlaceholder = 'Estimated end date (optional)';

beforeEach(() => {
  mockRouter = {
    push: jest.fn(), // the component uses `router.push` only
    pathname: '/onetime',
    asPath: '/onetime',
  };
  (useRouter as jest.Mock).mockReturnValue(mockRouter);
});

afterEach(() => {
  mockRouter = null;
  MockDate.reset();
});

describe('One time page', () => {
  let asFragment: any | null = null;
  let getByTestId: any | null = null;

  const renderComponent = () => {
    const view = render(
      <MuiPickersUtilsProvider utils={DayJSUtils}>
        <AppStateProvider initialStateOverride={initialStateOverrideWithQueryParamData}>
          <OneTimePage />
        </AppStateProvider>
      </MuiPickersUtilsProvider>,
      {}
    );
    ({ asFragment, getByTestId } = view);
  };

  beforeAll(() => {
    setAmPmFormat();
  });

  afterEach(() => {
    // cleanup on exiting
    asFragment = null;
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should enable next button after date selection', () => {
    renderComponent();
    const todayDate = dayjs();
    const startDate = todayDate.add(1, 'day').format('MM-DD-YYYY');
    const endDate = todayDate.add(3, 'day').format('MM-DD-YYYY');
    const startDateElement = screen.getByLabelText(startDatePlaceholder);
    const endDateElement = screen.getByLabelText(endDatePlaceholder);
    expect(screen.getByRole('button', { name: 'Next' })).toBeDisabled();

    fireEvent.change(startDateElement, { target: { value: startDate } });
    fireEvent.change(endDateElement, { target: { value: endDate } });

    expect(screen.getByRole('button', { name: 'Next' })).toBeEnabled();
  });

  it('should show the warning text if assigning past dates', () => {
    renderComponent();
    const todayDate = dayjs();
    const pastDate = todayDate.subtract(1, 'day').format('MM-DD-YYYY');
    const startDateElement = screen.getByLabelText(startDatePlaceholder);
    const endDateElement = screen.getByLabelText(endDatePlaceholder);

    fireEvent.change(startDateElement, { target: { value: pastDate } });
    fireEvent.change(endDateElement, { target: { value: pastDate } });

    expect(screen.getAllByText('Date should not be before minimal date').length).toBe(2);
  });

  it('ignores invalid date format and keeps digits', () => {
    renderComponent();
    const garbageStartDate = '10-10-202o0';
    const garbageEndDate = '10-11-202o0';
    const startDateElement = screen.getByLabelText(startDatePlaceholder) as HTMLInputElement;
    const endDateElement = screen.getByLabelText(endDatePlaceholder) as HTMLInputElement;

    fireEvent.change(startDateElement, { target: { value: garbageStartDate } });
    fireEvent.change(endDateElement, { target: { value: garbageEndDate } });

    expect(startDateElement.value).toEqual('10/10/2020');
    expect(endDateElement.value).toEqual('10/11/2020');
  });

  it('should show the warning text if assigning start date > 6 months', () => {
    renderComponent();
    const todayDate = dayjs();
    const futureDate = todayDate.add(6, 'month').add(1, 'day').format('MM-DD-YYYY');
    const startDateElement = screen.getByLabelText(startDatePlaceholder);

    fireEvent.change(startDateElement, { target: { value: futureDate } });

    expect(screen.getAllByText('Date should not be after maximal date').length).toBe(1);
  });

  it('should show the warning text if assigning end date > 1 year', () => {
    renderComponent();
    const todayDate = dayjs();
    const futureDate = todayDate.add(1, 'year').add(1, 'day').format('MM-DD-YYYY');
    const endDateElement = screen.getByLabelText(endDatePlaceholder);

    fireEvent.change(endDateElement, { target: { value: futureDate } });

    expect(screen.getAllByText('Date should not be after maximal date').length).toBe(1);
  });

  it('routes to the pay-for-care page after selection and button click', () => {
    renderComponent();
    const todayDate = dayjs();
    const startDate = todayDate.add(1, 'day').format('MM-DD-YYYY');
    const endDate = todayDate.add(3, 'day').format('MM-DD-YYYY');
    const startDateElement = screen.getByLabelText(startDatePlaceholder);
    const endDateElement = screen.getByLabelText(endDatePlaceholder);

    fireEvent.change(startDateElement, { target: { value: startDate } });
    fireEvent.change(endDateElement, { target: { value: endDate } });

    screen.getByRole('button', { name: 'Next' }).click();

    expect(mockRouter.push).toHaveBeenCalledWith(POST_A_JOB_ROUTES.PAY_FOR_CARE);
  });

  it('should change the text label when clicking on slider', () => {
    renderComponent();
    const sliderInput = getByTestId('time-slider');

    sliderInput.getBoundingClientRect = jest.fn(() => {
      return {
        bottom: 286.22918701171875,
        height: 28,
        left: 19.572917938232422,
        right: 583.0937919616699,
        top: 258.22918701171875,
        width: 563.5208740234375,
        x: 19.572917938232422,
        y: 258.22918701171875,
      };
    });

    expect(sliderInput).toBeInTheDocument();
    expect(screen.getByText('9:00 AM - 5:00 PM')).toBeInTheDocument();

    fireEvent.mouseDown(sliderInput, { clientX: 162, clientY: 302 });

    expect(screen.getByText('6:00 AM - 5:00 PM')).toBeInTheDocument();
  });

  it('should remain disabled the next button if date is empty', () => {
    renderComponent();
    const startDateElement = screen.getByLabelText(startDatePlaceholder);
    const endDateElement = screen.getByLabelText(endDatePlaceholder);

    fireEvent.change(startDateElement, { target: { value: '' } });
    fireEvent.change(endDateElement, { target: { value: '' } });

    expect(screen.getByRole('button', { name: 'Next' })).toBeDisabled();
  });

  it('should override start time when default start time is before earliest time', () => {
    renderComponent();
    // 11:08:50 in CST
    MockDate.set('2020-12-31T12:08:50');
    const startDateElement = screen.getByLabelText(startDatePlaceholder);
    expect(screen.getByText('9:00 AM - 5:00 PM')).toBeInTheDocument();
    fireEvent.change(startDateElement, { target: { value: '12/31/2020' } });
    expect(screen.getByText('1:00 PM - 5:00 PM')).toBeInTheDocument();
  });

  it('should use default times', () => {
    renderComponent();
    // 11:35:50 in CST
    MockDate.set('2021-01-01T12:35:50');
    const startDateElement = screen.getByLabelText(startDatePlaceholder);
    expect(screen.getByText('9:00 AM - 5:00 PM')).toBeInTheDocument();
    fireEvent.change(startDateElement, { target: { value: '01/02/2021' } });
    expect(screen.getByText('9:00 AM - 5:00 PM')).toBeInTheDocument();
  });

  it('should override both start and end time when before earliest time', () => {
    renderComponent();
    // 17:00:50 in CST
    MockDate.set('2020-12-31T18:00:50');
    const startDateElement = screen.getByLabelText(startDatePlaceholder);
    expect(screen.getByText('9:00 AM - 5:00 PM')).toBeInTheDocument();
    fireEvent.change(startDateElement, { target: { value: '12/31/2020' } });
    expect(screen.getByText('7:00 PM - 8:30 PM')).toBeInTheDocument();
  });

  it('should not allow you to select same day when after 20:00 PM', () => {
    renderComponent();
    // 20:35:50 in CST
    MockDate.set('2020-12-31T21:00:50');
    const startDateElement = screen.getByLabelText(startDatePlaceholder);
    expect(screen.getByText('9:00 AM - 5:00 PM')).toBeInTheDocument();
    fireEvent.change(startDateElement, { target: { value: '12/31/2020' } });
    expect(screen.getByText('9:00 AM - 5:00 PM')).toBeInTheDocument();
  });
});

describe('One time page -- bad state', () => {
  const { location } = window;

  beforeAll((): void => {
    // @ts-ignore
    delete window.location;
    // @ts-ignore
    window.location = {
      assign: jest.fn(),
    };
  });

  afterAll((): void => {
    window.location = location;
    mockRouter = null;
  });

  it('should render then redirect if default state is present', async () => {
    render(
      <MuiPickersUtilsProvider utils={DayJSUtils}>
        <AppStateProvider>
          <OneTimePage />
        </AppStateProvider>
      </MuiPickersUtilsProvider>
    );
    await waitFor(() => {
      expect(window.location.assign).toHaveBeenCalledWith('/');
    });
  });
});
