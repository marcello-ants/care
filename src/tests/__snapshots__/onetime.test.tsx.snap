// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`One time page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h2
        class="MuiTypography-root makeStyles-header-5 MuiTypography-h2"
      >
        When do you need care?
      </h2>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h4
        class="MuiTypography-root makeStyles-dateSubheader-2 MuiTypography-h4"
      >
        Dates
      </h4>
    </div>
    <div
      class="MuiGrid-root makeStyles-datePickerContainer-6 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-root-11"
        data-testid="keyboardDatePicker"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="startDate"
          id="startDate-label"
        >
          Estimated start date
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
            id="startDate"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-root-25"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17.79 4.49h-.47v-.94a.63.63 0 10-1.25 0v1H7.85v-1a.67.67 0 00-.62-.63.67.67 0 00-.63.63v1h-.47A3.13 3.13 0 003 7.7V18a3.14 3.14 0 003.13 3.13h11.74a2.83 2.83 0 002.19-.94A3.13 3.13 0 0021 18V7.62a3.27 3.27 0 00-3.21-3.13zM6.13 5.82h.47v.7a.68.68 0 00.63.63.68.68 0 00.62-.63v-.7h8.22v.7a.63.63 0 101.25 0v-.7h.47a1.86 1.86 0 011.8 1.72H4.25a1.93 1.93 0 011.88-1.72zm13 13.3a1.85 1.85 0 01-1.25.55H6.13a1.77 1.77 0 01-1.8-1.8V8.79h15.34v9.08a1.85 1.85 0 01-.55 1.25z"
                    fill="#182c42"
                  />
                  <path
                    d="M7.77 13.17a.86.86 0 000-1.72.86.86 0 000 1.72zM12 13.17a.86.86 0 10-.86-.86.85.85 0 00.86.86zM16.15 13.17a.86.86 0 10-.86-.86.86.86 0 00.86.86zM7.77 17.09a.86.86 0 10-.86-.86.86.86 0 00.86.86zM12 17.09a.86.86 0 10-.86-.86.86.86 0 00.86.86zM16.15 17.09a.86.86 0 000-1.72.86.86 0 000 1.72z"
                    fill="#182c42"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-datePickerContainer-6 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-root-11"
        data-testid="keyboardDatePicker"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="endDate"
          id="endDate-label"
        >
          Estimated end date (optional)
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
            id="endDate"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-root-26"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17.79 4.49h-.47v-.94a.63.63 0 10-1.25 0v1H7.85v-1a.67.67 0 00-.62-.63.67.67 0 00-.63.63v1h-.47A3.13 3.13 0 003 7.7V18a3.14 3.14 0 003.13 3.13h11.74a2.83 2.83 0 002.19-.94A3.13 3.13 0 0021 18V7.62a3.27 3.27 0 00-3.21-3.13zM6.13 5.82h.47v.7a.68.68 0 00.63.63.68.68 0 00.62-.63v-.7h8.22v.7a.63.63 0 101.25 0v-.7h.47a1.86 1.86 0 011.8 1.72H4.25a1.93 1.93 0 011.88-1.72zm13 13.3a1.85 1.85 0 01-1.25.55H6.13a1.77 1.77 0 01-1.8-1.8V8.79h15.34v9.08a1.85 1.85 0 01-.55 1.25z"
                    fill="#182c42"
                  />
                  <path
                    d="M7.77 13.17a.86.86 0 000-1.72.86.86 0 000 1.72zM12 13.17a.86.86 0 10-.86-.86.85.85 0 00.86.86zM16.15 13.17a.86.86 0 10-.86-.86.86.86 0 00.86.86zM7.77 17.09a.86.86 0 10-.86-.86.86.86 0 00.86.86zM12 17.09a.86.86 0 10-.86-.86.86.86 0 00.86.86zM16.15 17.09a.86.86 0 000-1.72.86.86 0 000 1.72z"
                    fill="#182c42"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h4
        class="MuiTypography-root makeStyles-timesSubheader-3 MuiTypography-h4"
      >
        Times
      </h4>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter"
          >
            <strong>
              9:00 AM - 5:00 PM
            </strong>
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <span
            class="MuiSlider-root MuiSlider-colorPrimary makeStyles-timeSlider-19"
            data-testid="time-slider"
          >
            <span
              class="MuiSlider-rail"
            />
            <span
              class="MuiSlider-track"
              style="left: 37.5%; width: 33.33333333333333%;"
            />
            <input
              type="hidden"
              value="9,17"
            />
            <span
              aria-orientation="horizontal"
              aria-valuemax="24"
              aria-valuemin="0"
              aria-valuenow="9"
              class="MuiSlider-thumb MuiSlider-thumbColorPrimary"
              data-index="0"
              role="slider"
              style="left: 37.5%;"
              tabindex="0"
            />
            <span
              aria-orientation="horizontal"
              aria-valuemax="24"
              aria-valuemin="0"
              aria-valuenow="17"
              class="MuiSlider-thumb MuiSlider-thumbColorPrimary"
              data-index="1"
              role="slider"
              style="left: 70.83333333333333%;"
              tabindex="0"
            />
          </span>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-nextButtonContainer-1 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge Mui-disabled MuiButton-fullWidth Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Next
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
